prov/shm/src/smr_msg.c

```diff
@@ -38,46 +38,51 @@
 #include "ofi_iov.h"
 #include "smr.h"
 
-
-static inline uint16_t smr_convert_rx_flags(uint64_t fi_flags)
+struct smr_rx_entry *smr_alloc_rx_entry(struct smr_srx_ctx *srx)
 {
-	uint16_t flags = 0;
+	if (ofi_freestack_isempty(srx->recv_fs)) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"not enough space to post recv\n");
+		return NULL;
+	}
 
-	if (fi_flags & FI_COMPLETION)
-		flags |= SMR_RX_COMPLETION;
-	if (fi_flags & FI_MULTI_RECV)
-		flags |= SMR_MULTI_RECV;
+	return ofi_freestack_pop(srx->recv_fs);
+}
 
-	return flags;
+void smr_init_rx_entry(struct smr_rx_entry *entry, const struct iovec *iov,
+		       void **desc, size_t count, fi_addr_t addr,
+		       void *context, uint64_t tag, uint64_t flags)
+{
+	memcpy(&entry->iov, iov, sizeof(*iov) * count);
+	if (desc)
+		memcpy(entry->desc, desc, sizeof(*desc) * count);
+
+	entry->peer_entry.iov = entry->iov;
+	entry->peer_entry.desc = entry->desc;
+	entry->peer_entry.count = count;
+	entry->peer_entry.addr = addr;
+	entry->peer_entry.context = context;
+	entry->peer_entry.tag = tag;
+	entry->peer_entry.flags = flags;
 }
 
-static struct smr_rx_entry *smr_get_recv_entry(struct smr_ep *ep,
+struct smr_rx_entry *smr_get_recv_entry(struct smr_ep *ep,
 		const struct iovec *iov, void **desc, size_t count, fi_addr_t addr,
 		void *context, uint64_t tag, uint64_t ignore, uint64_t flags)
 {
 	struct smr_rx_entry *entry;
 
-	if (ofi_freestack_isempty(ep->recv_fs)) {
-		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
-			"not enough space to post recv\n");
+	entry = smr_alloc_rx_entry(smr_get_srx(ep));
+	if (!entry)
 		return NULL;
-	}
 
-	entry = ofi_freestack_pop(ep->recv_fs);
+	smr_init_rx_entry(entry, iov, desc, count, addr,
+			  context, tag, flags);
 
-	memcpy(&entry->iov, iov, sizeof(*iov) * count);
-	entry->iov_count = count;
-	entry->context = context;
-	entry->err = 0;
-	entry->flags = smr_convert_rx_flags(flags);
-	entry->peer_id = ep->util_ep.caps & FI_DIRECTED_RECV &&
-				addr != FI_ADDR_UNSPEC ?
-				smr_addr_lookup(ep->util_ep.av, addr) : -1;
-	entry->tag = tag;
-	entry->ignore = ignore;
+	entry->peer_entry.owner_context = NULL;
 
-	entry->iface = smr_get_mr_hmem_iface(ep->util_ep.domain, desc,
-					     &entry->device);
+	ofi_atomic_initialize32(&entry->ref, 0);
```

Comment:
Initialize should only be called once for the structure.  This can end up abstracted as a spinlock around an integer.

Reply:
Do you prefer a single initialization or to abstract it with a lock?

Reply:
Just realized we always have the srx lock when modifying this value so I just dropped the protection completely